qcengine/programs/qchem.py

```diff
@@ -0,0 +1,251 @@
+
+        which_prog = which('qchem')
+        if which_prog not in self.version_cache:
+            with popen([which_prog, '-h'], popen_kwargs={"env": self._get_qc_path()}) as exc:
+                exc["proc"].wait(timeout=15)
+
+            if "QC not defined" in exc["stdout"]:
+                return safe_version("0.0.0")
+
+            self.version_cache[which_prog] = safe_version(exc["stdout"].splitlines()[0].split()[-1])
+
+        return self.version_cache[which_prog]
+
+    def compute(self, input_data: 'ResultInput', config: 'JobConfig') -> 'Result':
+        """
+        Run qchem
+        """
+        # Check if qchem executable is found
+        self.found(raise_error=True)
+
+        # Check qchem version
+        if parse_version(self.get_version()) < parse_version("5.2"):
+            raise TypeError("Q-Chem version '{}' not supported".format(self.get_version()))
+
+        # Setup the job
+        job_inputs = self.build_input(input_data, config)
+
+        # Run qchem
+        exe_success, proc = self.execute(job_inputs)
+
+        # Determine whether the calculation succeeded
+        if exe_success:
+            # If execution succeeded, collect results
+            result = self.parse_output(proc["outfiles"], input_data)
+            return result
+        else:
+            outfile = proc["outfiles"]["dispatch.out"]
+            if "fatal error occurred in module qparser" in outfile:
+                raise InputError(proc["outfiles"]["dispatch.out"])
+            else:
+                # Return UnknownError for error propagation
+                raise UnknownError(proc["outfiles"]["dispatch.out"])
+
+    def execute(self,
+                inputs: Dict[str, Any],
+                extra_infiles: Optional[Dict[str, str]] = None,
+                extra_outfiles: Optional[List[str]] = None,
+                extra_commands: Optional[List[str]] = None,
+                scratch_name: Optional[str] = None,
+                scratch_messy: bool = False,
+                timeout: Optional[int] = None) -> Tuple[bool, Dict[str, Any]]:
+        """
+        For option documentation go look at qcengine/util.execute
+        """
+
+        # Collect all input files and update with extra_infiles
+        infiles = inputs["infiles"]
+        if extra_infiles is not None:
+            infiles.update(extra_infiles)
+
+        binary_files = [os.path.join("savepath", x) for x in ["99.0", "131.0", "132.0"]]
+
+        # Collect all output files and extend with with extra_outfiles
+        outfiles = ["dispatch.out"]
+        if extra_outfiles is not None:
+            outfiles.extend(extra_outfiles)
+
+        # Replace commands with extra_commands if present
+        commands = inputs["commands"] + ["savepath"]
+        if extra_commands is not None:
+            commands = extra_commands
+
+        envs = self._get_qc_path()
+
+        with temporary_directory(parent=inputs["scratch_directory"], suffix="_qchem_scratch") as tmpdir:
```

Comment:
what new circle of hell hath q-chem cooked up to require managing tmp _outside_ `execute`?

Reply:
Something odd about where input files and scratch files need to live.